src/hazardous/secret_stream/xchacha20poly1305.rs

```diff
@@ -0,0 +1,455 @@
+		xor_buf8(self.get_inonce_mut(), &dst_out[macpos..macpos + 8]);
+		increment4(self.get_counter_mut());
+		if tag == Tag::REKEY
+			|| secure_cmp(
+				self.get_counter_mut(),
+				&[0u8; SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES],
+			)
+			.is_ok()
+		{
+			self.rekey()?;
+		}
+		Ok(())
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// TODO: Doc
+	/// Decryptes a message
+	pub fn pull(
+		&mut self,
+		cipher: &[u8],
+		ad: Option<&[u8]>,
+		plaintext_out: &mut [u8],
+		tag_out: &mut Tag,
+	) -> Result<(), UnknownCryptoError> {
+		if cipher.len() < SECRETSTREAM_XCHACHA20POLY1305_ABYTES {
+			return Err(UnknownCryptoError);
+		}
+		const BLOCKSIZE: usize = 64;
+		let mut block = [0u8; BLOCKSIZE];
+		let mut slen = [0u8; 8];
+		let pad = [0u8; 16];
+		let mlen = cipher.len() - SECRETSTREAM_XCHACHA20POLY1305_ABYTES;
+		let adlen = match ad {
+			Some(v) => v.len(),
+			None => 0,
+		};
+		let msgpos = core::mem::size_of::<Tag>();
+		let macpos = msgpos + mlen;
+		if plaintext_out.len() < mlen {
+			return Err(UnknownCryptoError);
+		}
+		chacha20_enc_in_place(&self.get_key(), &self.get_nonce(), 0, &mut block)?;
+		let mut poly = init(&OneTimeKey::from_slice(&block[..POLY1305_KEYSIZE])?);
+		use zeroize::Zeroize;
+		block.zeroize();
+		if adlen > 0 {
+			poly.update(ad.unwrap())?;
+		}
+		poly.update(&pad[..((16usize.overflowing_sub(adlen).0) & 15)])?;
+		block[0] = cipher[0];
+		chacha20_enc_in_place(&self.get_key(), &self.get_nonce(), 1, &mut block)?;
+		let tag = Tag::from_bits(block[0]).ok_or(UnknownCryptoError)?;
+		block[0] = cipher[0];
+		poly.update(&block)?;
+		poly.update(&cipher[msgpos..(msgpos + mlen)])?;
+		poly.update(&pad[..((16usize.overflowing_sub(BLOCKSIZE).0.overflowing_add(mlen).0) & 15)])?;
+		store_u64_into_le(&[adlen as u64], &mut slen);
+		poly.update(&slen)?;
+		store_u64_into_le(&[BLOCKSIZE as u64 + mlen as u64], &mut slen);
+		poly.update(&slen)?;
+		let mut mac = poly.finalize()?;
+		if !(mac == &cipher[macpos..macpos + mac.get_length()]) {
+			mac.zeroize();
+			return Err(UnknownCryptoError);
+		}
+		chacha20_enc(
+			&self.get_key(),
+			&self.get_nonce(),
+			2,
+			&cipher[msgpos..(msgpos + mlen)],
+			plaintext_out,
+		)?;
+		xor_buf8(self.get_inonce_mut(), mac.unprotected_as_bytes());
+		increment4(self.get_counter_mut());
+		if tag == Tag::REKEY
+			|| secure_cmp(
+				self.get_counter_mut(),
+				&[0u8; SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES],
+			)
+			.is_ok()
+		{
+			self.rekey()?;
+		}
+		*tag_out = tag;
+		Ok(())
+	}
+
+	/// Resets the internal message counter
+	fn reset_counter(&mut self) {
+		self.nonce[0] = 1;
+		self.nonce[1] = 0;
+		self.nonce[2] = 0;
+		self.nonce[3] = 0;
+	}
+}
+
+#[cfg(test)]
+mod public {}
+
+#[cfg(test)]
+mod private {
+	extern crate hex;
+	use self::hex::decode;
+	use super::*;
+	use hex::encode;
+
+	//All test values generated with libsodium https://github.com/jedisct1/libsodium version 1.0.18
+
+	#[test]
```

Comment:
This is great! It would be awesome if you could include the script/program that was used to create the test-vectors in `/tests/test_generation`.

Reply:
Fixed in https://github.com/brycx/orion/pull/99/commits/c962e9b5124307bf108a4dfb3b6a858128ca24f3